### DISCUSSION TITLE
clangd cannt find iostream and other stander library

### DISCUSSION THREAD

**Discussion by RI-ss** [clangd/clangd#1510] [Q&A] [answered]
case 1
![image](https://user-images.githubusercontent.com/71321125/219839600-acff9f7a-3f31-4526-b2d6-6eda0b750f47.png)
setting is such like this
comilpe_commands.json
![image](https://user-images.githubusercontent.com/71321125/219839828-52569c85-82fc-4fbb-8d6a-ea989879dded.png)
setting.json
![image](https://user-images.githubusercontent.com/71321125/219840067-a73bd26a-8c65-47dc-927c-0573f2a6df74.png)

it shows cannt find iostream but can find sdl2

case 2 but when i delete the comilpe_commands.json
![image](https://user-images.githubusercontent.com/71321125/219840346-826f5ea3-3498-44cc-81e6-0a104be685ad.png)
![image](https://user-images.githubusercontent.com/71321125/219841330-e5e55039-3ff0-4830-9e4f-2beddfd2f8ca.png)
![image](https://user-images.githubusercontent.com/71321125/219841759-fedb44e1-53aa-45b2-a767-4900ed1d89bc.png)
if can find the iosteam and stdio.h but cannt find sdl2
so what i need  to do to let it finds iostream and sdl2

**Comment by HighCommander4** [comment, marked as answer]
The argument to `--query-driver` needs to be the compiler executable.

So, if you change the argument to `--query-driver` from `C:\\msys64\\ming64\\include\\c++\\12.2.0` to `C:\\msys64\\mingw64\\bin\\clang-cl.exe`, and keep the `compile_commands.json`, things should work.

**Reply by RI-ss** [reply]
![image](https://user-images.githubusercontent.com/71321125/219843151-71f6af8d-1410-434d-a605-226a939f5969.png)
![image](https://user-images.githubusercontent.com/71321125/219843166-3ef6237a-126c-44af-827b-fbabe5ba1021.png)
i had changed it and restart the server but it still not work

**Reply by HighCommander4** [reply]
Can you show [logs](https://clangd.llvm.org/troubleshooting#gathering-logs) please?

**Reply by RI-ss** [reply]
is this logs?

V[14:02:10.011] <<< {"id":6,"jsonrpc":"2.0","method":"textDocument/codeAction","params":{"context":{"diagnostics":[{"code":"pp_file_not_found","message":"'cstdio' file not found","range":{"end":{"character":17,"line":3},"start":{"character":9,"line":3}},"relatedInformation":[],"severity":1,"source":"clang"}],"triggerKind":2},"range":{"end":{"character":17,"line":3},"start":{"character":17,"line":3}},"textDocument":{"uri":"file:///d%3A/c%2B%2B/test1/src/main.cpp"}}}

I[14:02:10.011] <-- textDocument/codeAction(6)
V[14:02:10.015] ASTWorker running EnumerateTweaks on version 33 of d:/c++/test1/src/main.cpp
I[14:02:10.015] --> reply:textDocument/codeAction(6) 3 ms
V[14:02:10.015] >>> {"id":6,"jsonrpc":"2.0","result":[]}

I[14:02:10.015] --> textDocument/clangd.fileStatus
V[14:02:10.015] >>> {"jsonrpc":"2.0","method":"textDocument/clangd.fileStatus","params":{"state":"idle","uri":"file:///d:/c%2B%2B/test1/src/main.cpp"}}

V[14:02:11.196] <<< {"id":7,"jsonrpc":"2.0","method":"shutdown"}

I[14:02:11.196] <-- shutdown(7)
I[14:02:11.196] --> reply:shutdown(7) 0 ms
V[14:02:11.196] >>> {"id":7,"jsonrpc":"2.0","result":null}

V[14:02:11.221] <<< {"jsonrpc":"2.0","method":"exit"}

I[14:02:11.221] <-- exit
I[14:02:11.221] LSP finished, exiting with status 0
The flag `-clang-tidy-checks` is obsolete and ignored.
I[14:02:11.271] clangd version 15.0.7
I[14:02:11.273] Features: windows
I[14:02:11.273] PID: 22864
I[14:02:11.273] Working directory: d:/c++/test1
I[14:02:11.273] argv[0]: C:\msys64\mingw64\bin\clangd.exe
I[14:02:11.273] argv[1]: --background-index
I[14:02:11.273] argv[2]: --compile-commands-dir=.
I[14:02:11.273] argv[3]: -j=13
I[14:02:11.273] argv[4]: --clang-tidy
I[14:02:11.273] argv[5]: --clang-tidy-checks=performance-*,bugprone-*
I[14:02:11.273] argv[6]: --all-scopes-completion
I[14:02:11.273] argv[7]: --completion-style=detailed
I[14:02:11.273] argv[8]: --header-insertion=iwyu
I[14:02:11.273] argv[9]: --pch-storage=memory
I[14:02:11.273] argv[10]: --log=verbose
I[14:02:11.273] argv[11]: --header-insertion=never
I[14:02:11.273] argv[12]: --query-driver=C:\msys64\mingw64\bin\clang-cl.exe
V[14:02:11.276] User config file is C:/Users/lizex/AppData/Local/clangd/config.yaml
I[14:02:11.276] Starting LSP over stdin/stdout
V[14:02:11.277] <<< {"id":0,"jsonrpc":"2.0","method":"initialize","params":{"capabilities":{"general":{"markdown":{"parser":"marked","version":"1.1.0"},"positionEncodings":["utf-16"],"regularExpressions":{"engine":"ECMAScript","version":"ES2020"},"staleRequestSupport":{"cancel":true,"retryOnContentModified":["textDocument/semanticTokens/full","textDocument/semanticTokens/range","textDocument/semanticTokens/full/delta"]}},"notebookDocument":{"synchronization":{"dynamicRegistration":true,"executionSummarySupport":true}},"textDocument":{"callHierarchy":{"dynamicRegistration":true},"codeAction":{"codeActionLiteralSupport":{"codeActionKind":{"valueSet":["","quickfix","refactor","refactor.extract","refactor.inline","refactor.rewrite","source","source.organizeImports"]}},"dataSupport":true,"disabledSupport":true,"dynamicRegistration":true,"honorsChangeAnnotations":false,"isPreferredSupport":true,"resolveSupport":{"properties":["edit"]}},"codeLens":{"dynamicRegistration":true},"colorProvider":{"dynamicRegistration":true},"completion":{"completionItem":{"commitCharactersSupport":true,"deprecatedSupport":true,"documentationFormat":["markdown","plaintext"],"insertReplaceSupport":true,"insertTextModeSupport":{"valueSet":[1,2]},"labelDetailsSupport":true,"preselectSupport":true,"resolveSupport":{"properties":["documentation","detail","additionalTextEdits"]},"snippetSupport":true,"tagSupport":{"valueSet":[1]}},"completionItemKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25]},"completionList":{"itemDefaults":["commitCharacters","editRange","insertTextFormat","insertTextMode"]},"contextSupport":true,"dynamicRegistration":true,"editsNearCursor":true,"insertTextMode":2},"declaration":{"dynamicRegistration":true,"linkSupport":true},"definition":{"dynamicRegistration":true,"linkSupport":true},"diagnostic":{"dynamicRegistration":true,"relatedDocumentSupport":false},"documentHighlight":{"dynamicRegistration":true},"documentLink":{"dynamicRegistration":true,"tooltipSupport":true},"documentSymbol":{"dynamicRegistration":true,"hierarchicalDocumentSymbolSupport":true,"labelSupport":true,"symbolKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26]},"tagSupport":{"valueSet":[1]}},"foldingRange":{"dynamicRegistration":true,"foldingRange":{"collapsedText":false},"foldingRangeKind":{"valueSet":["comment","imports","region"]},"lineFoldingOnly":true,"rangeLimit":5000},"formatting":{"dynamicRegistration":true},"hover":{"contentFormat":["markdown","plaintext"],"dynamicRegistration":true},"implementation":{"dynamicRegistration":true,"linkSupport":true},"inlayHint":{"dynamicRegistration":true,"resolveSupport":{"properties":["tooltip","textEdits","label.tooltip","label.location","label.command"]}},"inlineValue":{"dynamicRegistration":true},"linkedEditingRange":{"dynamicRegistration":true},"onTypeFormatting":{"dynamicRegistration":true},"publishDiagnostics":{"codeDescriptionSupport":true,"dataSupport":true,"relatedInformation":true,"tagSupport":{"valueSet":[1,2]},"versionSupport":false},"rangeFormatting":{"dynamicRegistration":true},"references":{"dynamicRegistration":true},"rename":{"dynamicRegistration":true,"honorsChangeAnnotations":true,"prepareSupport":true,"prepareSupportDefaultBehavior":1},"selectionRange":{"dynamicRegistration":true},"semanticTokens":{"augmentsSyntaxTokens":true,"dynamicRegistration":true,"formats":["relative"],"multilineTokenSupport":false,"overlappingTokenSupport":false,"requests":{"full":{"delta":true},"range":true},"serverCancelSupport":true,"tokenModifiers":["declaration","definition","readonly","static","deprecated","abstract","async","modification","documentation","defaultLibrary"],"tokenTypes":["namespace","type","class","enum","interface","struct","typeParameter","parameter","variable","property","enumMember","event","function","method","macro","keyword","modifier","comment","string","number","regexp","operator","decorator"]},"signatureHelp":{"contextSupport":true,"dynamicRegistration":true,"signatureInformation":{"activeParameterSupport":true,"documentationFormat":["markdown","plaintext"],"parameterInformation":{"labelOffsetSupport":true}}},"synchronization":{"didSave":true,"dynamicRegistration":true,"willSave":true,"willSaveWaitUntil":true},"typeDefinition":{"dynamicRegistration":true,"linkSupport":true},"typeHierarchy":{"dynamicRegistration":true}},"window":{"showDocument":{"support":true},"showMessage":{"messageActionItem":{"additionalPropertiesSupport":true}},"workDoneProgress":true},"workspace":{"applyEdit":true,"codeLens":{"refreshSupport":true},"configuration":true,"diagnostics":{"refreshSupport":true},"didChangeConfiguration":{"dynamicRegistration":true},"didChangeWatchedFiles":{"dynamicRegistration":true,"relativePatternSupport":true},"executeCommand":{"dynamicRegistration":true},"fileOperations":{"didCreate":true,"didDelete":true,"didRename":true,"dynamicRegistration":true,"willCreate":true,"willDelete":true,"willRename":true},"inlayHint":{"refreshSupport":true},"inlineValue":{"refreshSupport":true},"semanticTokens":{"refreshSupport":true},"symbol":{"dynamicRegistration":true,"resolveSupport":{"properties":["location.range"]},"symbolKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26]},"tagSupport":{"valueSet":[1]}},"workspaceEdit":{"changeAnnotationSupport":{"groupsOnLabel":true},"documentChanges":true,"failureHandling":"textOnlyTransactional","normalizesLineEndings":true,"resourceOperations":["create","rename","delete"]},"workspaceFolders":true}},"clientInfo":{"name":"Visual Studio Code","version":"1.75.1"},"initializationOptions":{"clangdFileStatus":true,"fallbackFlags":["-std=c++20","-IC:\\msys64\\mingw64\\include\\c++\\12.2.0"]},"locale":"zh-cn","processId":2428,"rootPath":"d:\\c++\\test1","rootUri":"file:///d%3A/c%2B%2B/test1","trace":"off","workspaceFolders":[{"name":"test1","uri":"file:///d%3A/c%2B%2B/test1"}]}}

I[14:02:11.277] <-- initialize(0)
I[14:02:11.279] --> reply:initialize(0) 2 ms
V[14:02:11.279] >>> {"id":0,"jsonrpc":"2.0","result":{"capabilities":{"astProvider":true,"callHierarchyProvider":true,"clangdInlayHintsProvider":true,"codeActionProvider":{"codeActionKinds":["quickfix","refactor","info"]},"compilationDatabase":{"automaticReload":true},"completionProvider":{"resolveProvider":false,"triggerCharacters":[".","<",">",":","\"","/","*"]},"declarationProvider":true,"definitionProvider":true,"documentFormattingProvider":true,"documentHighlightProvider":true,"documentLinkProvider":{"resolveProvider":false},"documentOnTypeFormattingProvider":{"firstTriggerCharacter":"\n","moreTriggerCharacter":[]},"documentRangeFormattingProvider":true,"documentSymbolProvider":true,"executeCommandProvider":{"commands":["clangd.applyFix","clangd.applyTweak"]},"hoverProvider":true,"implementationProvider":true,"inlayHintProvider":true,"memoryUsageProvider":true,"referencesProvider":true,"renameProvider":{"prepareProvider":true},"selectionRangeProvider":true,"semanticTokensProvider":{"full":{"delta":true},"legend":{"tokenModifiers":["declaration","deprecated","deduced","readonly","static","abstract","virtual","dependentName","defaultLibrary","usedAsMutableReference","functionScope","classScope","fileScope","globalScope"],"tokenTypes":["variable","variable","parameter","function","method","function","property","variable","class","interface","enum","enumMember","type","type","unknown","namespace","typeParameter","concept","type","macro","comment"]},"range":false},"signatureHelpProvider":{"triggerCharacters":["(",")","{","}","<",">",","]},"standardTypeHierarchyProvider":true,"textDocumentSync":{"change":2,"openClose":true,"save":true},"typeDefinitionProvider":true,"typeHierarchyProvider":true,"workspaceSymbolProvider":true},"serverInfo":{"name":"clangd","version":"clangd version 15.0.7 windows x86_64-w64-windows-gnu"}}}

V[14:02:11.281] <<< {"jsonrpc":"2.0","method":"initialized","params":{}}

I[14:02:11.282] <-- initialized
V[14:02:11.284] <<< {"jsonrpc":"2.0","method":"textDocument/didOpen","params":{"textDocument":{"languageId":"cpp","text":"#include <iostream>\r\nusing namespace std;\r\n\r\n#include <cstdio>\r\n#include <cstring>\r\n#include <stdio.h>\r\n#include \"SDL2/SDL.h\"\r\n\r\n// #include \"SDL2/SDL.h\"\r\n#define ull unsigned long long\r\n#define ll long long\r\n#define Max(a, b) a > b ? a : b\r\nusing ULL = unsigned long long;\r\n\r\nclass mod {\r\npublic:\r\n  int a;\r\n  long long b;\r\n  ULL shi;\r\n\r\n  static mod New() { return mod{}; }\r\n\r\n  mod const &operator=(mod &value) {\r\n    a = value.a;\r\n    b = value.b;\r\n    cout << \"move\" << endl;\r\n    return value;\r\n  }\r\n\r\n  void operator=(mod &&value) noexcept {}\r\n\r\n  void operator=(mod const &value) {}\r\n\r\n  mod(const mod &value) {\r\n    a = value.a;\r\n    b = value.b;\r\n  }\r\n\r\nprivate:\r\n  mod() {}\r\n};\r\n\r\nstruct particle {\r\n  int a;\r\n  long long b;\r\n  char c[10];\r\n};\r\n\r\nparticle shitting(int a, long long b) { return particle{a, b}; }\r\n\r\ntypedef particle (*c)(int, long long);\r\n\r\nvoid get(particle (*c)(int, long long)) { c(1, 2); }\r\n\r\nint wmain(int argc, char **argv) {\r\n  cout << \"shit\" << endl;\r\n  auto m = mod::New();\r\n\r\n  auto shi = 2ull;\r\n  m.a = 1;\r\n  m.b = 2;\r\n  auto m2 = m;\r\n\r\n  m2 = mod::New();\r\n\r\n  auto str = \"1234\";\r\n\r\n  auto len = strlen(str);\r\n\r\n  cout << sizeof(particle) << endl;\r\n\r\n  particle p = {.a = 1, .b = 2};\r\n\r\n  int a = 1, b = 2;\r\n  double c = a / 2.0;\r\n  particle (*noshit)(int, long long) = shitting;\r\n\r\n  int ends = (1, 2, 3);\r\n  return 0;\r\n}\r\n","uri":"file:///d%3A/c%2B%2B/test1/src/main.cpp","version":33}}}

I[14:02:11.284] <-- textDocument/didOpen
I[14:02:11.288] Loaded compilation database from d:/c++/test1/./compile_commands.json
V[14:02:11.288] Broadcasting compilation database from d:/c++/test1/.
I[14:02:11.289] --> window/workDoneProgress/create(0)
V[14:02:11.289] >>> {"id":0,"jsonrpc":"2.0","method":"window/workDoneProgress/create","params":{"token":"backgroundIndexProgress"}}

I[14:02:11.290] Enqueueing 2 commands for indexing
V[14:02:11.290] <<< {"id":0,"jsonrpc":"2.0","result":null}

I[14:02:11.290] <-- reply(0)
I[14:02:11.290] --> $/progress
V[14:02:11.290] >>> {"jsonrpc":"2.0","method":"$/progress","params":{"token":"backgroundIndexProgress","value":{"kind":"begin","percentage":0,"title":"indexing"}}}

I[14:02:11.290] --> $/progress
V[14:02:11.290] >>> {"jsonrpc":"2.0","method":"$/progress","params":{"token":"backgroundIndexProgress","value":{"kind":"report","message":"0/1","percentage":0}}}

V[14:02:11.301] <<< {"id":1,"jsonrpc":"2.0","method":"textDocument/documentSymbol","params":{"textDocument":{"uri":"file:///d%3A/c%2B%2B/test1/src/main.cpp"}}}

I[14:02:11.301] <-- textDocument/documentSymbol(1)
V[14:02:11.302] <<< {"id":2,"jsonrpc":"2.0","method":"textDocument/codeAction","params":{"context":{"diagnostics":[],"triggerKind":2},"range":{"end":{"character":17,"line":3},"start":{"character":17,"line":3}},"textDocument":{"uri":"file:///d%3A/c%2B%2B/test1/src/main.cpp"}}}

I[14:02:11.302] <-- textDocument/codeAction(2)
V[14:02:11.317] <<< {"id":3,"jsonrpc":"2.0","method":"textDocument/documentLink","params":{"textDocument":{"uri":"file:///d%3A/c%2B%2B/test1/src/main.cpp"}}}

I[14:02:11.317] <-- textDocument/documentLink(3)
V[14:02:11.317] <<< {"id":4,"jsonrpc":"2.0","method":"textDocument/inlayHint","params":{"range":{"end":{"character":25,"line":55},"start":{"character":0,"line":0}},"textDocument":{"uri":"file:///d%3A/c%2B%2B/test1/src/main.cpp"}}}

I[14:02:11.317] <-- textDocument/inlayHint(4)
E[14:02:11.333] System include extraction: driver execution failed with return code: 1 - ''. Args: ["C:\\msys64\\mingw64\\bin\\clang-cl.exe" -E -x c++ - -v]
I[14:02:11.334] ASTWorker building file d:/c++/test1/src/main.cpp version 33 with command 
[D:\c++\test1]
C:/msys64/mingw64/bin/clang-cl.exe --driver-mode=cl -c /EHsc -Zi -FS "-Fdd:\\c++\\test1\\build\\windows\\x64\\debug\\compile.test1.pdb" -Od -std:c++20 "-ID:\\Program Files\\vcpkg\\installed\\x64-windows-static\\include" "-Fod:\\c++\\test1\\build\\.objs\\test1\\windows\\x64\\debug\\src\\main.cpp.obj" -resource-dir=C:/msys64/mingw64/lib/clang/15.0.7 -- d:/c++/test1/src/main.cpp
V[14:02:11.338] Driver produced command: cc1 -cc1 -triple x86_64-pc-windows-msvc19.20.0 -fsyntax-only -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name main.cpp -mrelocation-model pic -pic-level 2 -mframe-pointer=none -relaxed-aliasing -fmath-errno -ffp-contract=on -fno-rounding-math -mconstructor-aliases -funwind-tables=2 -target-cpu x86-64 -mllvm -x86-asm-syntax=intel -tune-cpu generic -mllvm -treat-scalable-fixed-error-as-warning -D_MT -flto-visibility-public-std --dependent-lib=libcmt --dependent-lib=oldnames -stack-protector 2 -fcxx-exceptions -fexceptions -fexternc-nounwind -fms-volatile -fdiagnostics-format msvc -gno-column-info -gcodeview -debug-info-kind=constructor "-fcoverage-compilation-dir=D:\\c++\\test1" -resource-dir C:/msys64/mingw64/lib/clang/15.0.7 -I "D:\\Program Files\\vcpkg\\installed\\x64-windows-static\\include" -internal-isystem C:/msys64/mingw64/lib/clang/15.0.7/include -internal-isystem "D:\\Program Files\\vcpkg\\installed\\x64-windows\\include" -O0 -fdeprecated-macro "-fdebug-compilation-dir=D:\\c++\\test1" -ferror-limit 19 -fno-use-cxa-atexit -fms-extensions -fms-compatibility -fms-compatibility-version=19.20 -std=c++20 -fdelayed-template-parsing -fno-implicit-modules -no-round-trip-args -faddrsig -x c++ d:/c++/test1/src/main.cpp
I[14:02:11.338] --> textDocument/clangd.fileStatus
V[14:02:11.338] >>> {"jsonrpc":"2.0","method":"textDocument/clangd.fileStatus","params":{"state":"parsing includes, running Update","uri":"file:///d:/c%2B%2B/test1/src/main.cpp"}}

V[14:02:11.338] Building first preamble for d:/c++/test1/src/main.cpp version 33
V[14:02:11.346] indexed preamble AST for d:/c++/test1/src/main.cpp version 33:
  symbol slab: 1 symbols, 4448 bytes
  ref slab: 0 symbols, 0 refs, 128 bytes
  relations slab: 0 relations, 24 bytes
V[14:02:11.346] Build dynamic index for header symbols with estimated memory usage of 6828 bytes
V[14:02:11.347] Built preamble of size 218208 for file d:/c++/test1/src/main.cpp version 33 in 0.01 seconds
I[14:02:11.347] --> workspace/semanticTokens/refresh(1)
V[14:02:11.347] >>> {"id":1,"jsonrpc":"2.0","method":"workspace/semanticTokens/refresh","params":null}

I[14:02:11.347] --> textDocument/clangd.fileStatus
V[14:02:11.347] >>> {"jsonrpc":"2.0","method":"textDocument/clangd.fileStatus","params":{"state":"parsing includes, running Build AST","uri":"file:///d:/c%2B%2B/test1/src/main.cpp"}}

I[14:02:11.347] --> textDocument/clangd.fileStatus
V[14:02:11.347] >>> {"jsonrpc":"2.0","method":"textDocument/clangd.fileStatus","params":{"state":"parsing includes, parsing main file","uri":"file:///d:/c%2B%2B/test1/src/main.cpp"}}

V[14:02:11.349] <<< {"id":1,"jsonrpc":"2.0","result":null}

I[14:02:11.349] <-- reply(1)
V[14:02:11.367] BackgroundIndex: building version 1 after loading index from disk
V[14:02:11.457] BackgroundIndex: serving version 1 (11050752 bytes)
I[14:02:11.483] --> $/progress
V[14:02:11.483] >>> {"jsonrpc":"2.0","method":"$/progress","params":{"token":"backgroundIndexProgress","value":{"kind":"report","message":"0/1","percentage":0}}}

I[14:02:11.484] --> $/progress
V[14:02:11.484] >>> {"jsonrpc":"2.0","method":"$/progress","params":{"token":"backgroundIndexProgress","value":{"kind":"end"}}}

V[14:02:11.805] indexed file AST for d:/c++/test1/src/main.cpp version 33:
  symbol slab: 22 symbols, 9320 bytes
  ref slab: 22 symbols, 58 refs, 4992 bytes
  relations slab: 0 relations, 24 bytes
V[14:02:11.805] Build dynamic index for main-file symbols with estimated memory usage of 23832 bytes
I[14:02:11.805] --> textDocument/publishDiagnostics
V[14:02:11.805] >>> {"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"diagnostics":[{"code":"pp_file_not_found","message":"'iostream' file not found","range":{"end":{"character":19,"line":0},"start":{"character":9,"line":0}},"relatedInformation":[],"severity":1,"source":"clang"},{"code":"pp_file_not_found","message":"'cstdio' file not found","range":{"end":{"character":17,"line":3},"start":{"character":9,"line":3}},"relatedInformation":[],"severity":1,"source":"clang"}],"uri":"file:///d:/c%2B%2B/test1/src/main.cpp","version":33}}

V[14:02:11.806] ASTWorker running DocumentSymbols on version 33 of d:/c++/test1/src/main.cpp
I[14:02:11.806] --> reply:textDocument/documentSymbol(1) 505 ms
V[14:02:11.807] >>> {"id":1,"jsonrpc":"2.0","result":[{"kind":13,"name":"using namespace std","range":{"end":{"character":19,"line":1},"start":{"character":0,"line":1}},"selectionRange":{"end":{"character":19,"line":1},"start":{"character":16,"line":1}}},{"detail":"type alias","kind":5,"name":"ULL","range":{"end":{"character":30,"line":12},"start":{"character":0,"line":12}},"selectionRange":{"end":{"character":9,"line":12},"start":{"character":6,"line":12}}},{"children":[{"detail":"int","kind":8,"name":"a","range":{"end":{"character":7,"line":16},"start":{"character":2,"line":16}},"selectionRange":{"end":{"character":7,"line":16},"start":{"character":6,"line":16}}},{"detail":"long long","kind":8,"name":"b","range":{"end":{"character":13,"line":17},"start":{"character":2,"line":17}},"selectionRange":{"end":{"character":13,"line":17},"start":{"character":12,"line":17}}},{"detail":"ULL","kind":8,"name":"shi","range":{"end":{"character":9,"line":18},"start":{"character":2,"line":18}},"selectionRange":{"end":{"character":9,"line":18},"start":{"character":6,"line":18}}},{"detail":"mod ()","kind":6,"name":"New","range":{"end":{"character":36,"line":20},"start":{"character":2,"line":20}},"selectionRange":{"end":{"character":16,"line":20},"start":{"character":13,"line":20}}},{"detail":"const mod &(mod &)","kind":6,"name":"operator=","range":{"end":{"character":3,"line":27},"start":{"character":2,"line":22}},"selectionRange":{"end":{"character":21,"line":22},"start":{"character":13,"line":22}}},{"detail":"void (mod &&) noexcept","kind":6,"name":"operator=","range":{"end":{"character":41,"line":29},"start":{"character":2,"line":29}},"selectionRange":{"end":{"character":15,"line":29},"start":{"character":7,"line":29}}},{"detail":"void (const mod &)","kind":6,"name":"operator=","range":{"end":{"character":37,"line":31},"start":{"character":2,"line":31}},"selectionRange":{"end":{"character":15,"line":31},"start":{"character":7,"line":31}}},{"detail":"(const mod &)","kind":9,"name":"mod","range":{"end":{"character":3,"line":36},"start":{"character":2,"line":33}},"selectionRange":{"end":{"character":5,"line":33},"start":{"character":2,"line":33}}},{"detail":"()","kind":9,"name":"mod","range":{"end":{"character":10,"line":39},"start":{"character":2,"line":39}},"selectionRange":{"end":{"character":5,"line":39},"start":{"character":2,"line":39}}}],"detail":"class","kind":5,"name":"mod","range":{"end":{"character":1,"line":40},"start":{"character":0,"line":14}},"selectionRange":{"end":{"character":9,"line":14},"start":{"character":6,"line":14}}},{"children":[{"detail":"int","kind":8,"name":"a","range":{"end":{"character":7,"line":43},"start":{"character":2,"line":43}},"selectionRange":{"end":{"character":7,"line":43},"start":{"character":6,"line":43}}},{"detail":"long long","kind":8,"name":"b","range":{"end":{"character":13,"line":44},"start":{"character":2,"line":44}},"selectionRange":{"end":{"character":13,"line":44},"start":{"character":12,"line":44}}},{"detail":"char[10]","kind":8,"name":"c","range":{"end":{"character":12,"line":45},"start":{"character":2,"line":45}},"selectionRange":{"end":{"character":8,"line":45},"start":{"character":7,"line":45}}}],"detail":"struct","kind":23,"name":"particle","range":{"end":{"character":1,"line":46},"start":{"character":0,"line":42}},"selectionRange":{"end":{"character":15,"line":42},"start":{"character":7,"line":42}}},{"detail":"particle (int, long long)","kind":12,"name":"shitting","range":{"end":{"character":64,"line":48},"start":{"character":0,"line":48}},"selectionRange":{"end":{"character":17,"line":48},"start":{"character":9,"line":48}}},{"detail":"type alias","kind":5,"name":"c","range":{"end":{"character":37,"line":50},"start":{"character":0,"line":50}},"selectionRange":{"end":{"character":20,"line":50},"start":{"character":19,"line":50}}},{"detail":"void (particle (*)(int, long long))","kind":12,"name":"get","range":{"end":{"character":52,"line":52},"start":{"character":0,"line":52}},"selectionRange":{"end":{"character":8,"line":52},"start":{"character":5,"line":52}}},{"detail":"int (int, char **)","kind":12,"name":"wmain","range":{"end":{"character":1,"line":79},"start":{"character":0,"line":54}},"selectionRange":{"end":{"character":9,"line":54},"start":{"character":4,"line":54}}}]}

V[14:02:11.807] ASTWorker running EnumerateTweaks on version 33 of d:/c++/test1/src/main.cpp
I[14:02:11.807] --> reply:textDocument/codeAction(2) 504 ms
V[14:02:11.807] >>> {"id":2,"jsonrpc":"2.0","result":[]}

V[14:02:11.807] ASTWorker running DocumentLinks on version 33 of d:/c++/test1/src/main.cpp
I[14:02:11.807] --> reply:textDocument/documentLink(3) 490 ms
V[14:02:11.807] >>> {"id":3,"jsonrpc":"2.0","result":[{"range":{"end":{"character":21,"line":6},"start":{"character":9,"line":6}},"target":"file:///D:/Program%20Files/vcpkg/installed/x64-windows-static/include/SDL2/SDL.h"}]}

V[14:02:11.807] ASTWorker running InlayHints on version 33 of d:/c++/test1/src/main.cpp
I[14:02:11.807] --> reply:textDocument/inlayHint(4) 489 ms
V[14:02:11.807] >>> {"id":4,"jsonrpc":"2.0","result":[{"label":".a=","paddingLeft":false,"paddingRight":false,"position":{"character":56,"line":48}},{"label":".b=","paddingLeft":false,"paddingRight":false,"position":{"character":59,"line":48}}]}

I[14:02:11.807] --> textDocument/clangd.fileStatus
V[14:02:11.807] >>> {"jsonrpc":"2.0","method":"textDocument/clangd.fileStatus","params":{"state":"idle","uri":"file:///d:/c%2B%2B/test1/src/main.cpp"}}

V[14:02:12.077] <<< {"id":5,"jsonrpc":"2.0","method":"textDocument/codeAction","params":{"context":{"diagnostics":[{"code":"pp_file_not_found","message":"'cstdio' file not found","range":{"end":{"character":17,"line":3},"start":{"character":9,"line":3}},"relatedInformation":[],"severity":1,"source":"clang"}],"triggerKind":2},"range":{"end":{"character":17,"line":3},"start":{"character":17,"line":3}},"textDocument":{"uri":"file:///d%3A/c%2B%2B/test1/src/main.cpp"}}}

I[14:02:12.077] <-- textDocument/codeAction(5)
V[14:02:12.077] ASTWorker running EnumerateTweaks on version 33 of d:/c++/test1/src/main.cpp
I[14:02:12.077] --> reply:textDocument/codeAction(5) 0 ms
V[14:02:12.077] >>> {"id":5,"jsonrpc":"2.0","result":[]}

I[14:02:12.078] --> textDocument/clangd.fileStatus
V[14:02:12.078] >>> {"jsonrpc":"2.0","method":"textDocument/clangd.fileStatus","params":{"state":"idle","uri":"file:///d:/c%2B%2B/test1/src/main.cpp"}}

**Reply by HighCommander4** [reply]
Yes, these are the logs I was looking for.

This line is interesting:

> E[14:02:11.333] System include extraction: driver execution failed with return code: 1 - ''. Args: ["C:\msys64\mingw64\bin\clang-cl.exe" -E -x c++ - -v]

It suggests that clangd's mechanism for discovering the built-in include paths of the compiler is failing for some reason. It's not clear why.

Could you try running the command `"C:\msys64\mingw64\bin\clang-cl.exe" -E -x c++ - -v` in a terminal and see what happens?

**Reply by RI-ss** [reply]
![image](https://user-images.githubusercontent.com/71321125/219844426-9d015c9d-3a50-4435-82c7-3f79d9ac3a8a.png)

**Reply by HighCommander4** [reply]
Hmm, not quite sure what to make of that.

Perhaps someone more familiar with Windows came chime in on this.

**Reply by RI-ss** [reply]
![image](https://user-images.githubusercontent.com/71321125/219848214-3e93381a-94e9-4535-8f90-ca16863ffcd6.png)
maybe this is helpful

**Reply by HighCommander4** [reply]
Interesting, maybe `clang-cl.exe` has a different command line syntax and would require special support in `SystemIncludeExtractor`? (But if so, then I'm surprised no one has run into this before...)

Does it work if you change `clang-cl.exe` to `clang.exe`, in both `compile_commands.json` and the argument to `--query-driver`? You may need to remove the `/EHsc` line from `compile_commands.json` as well.

**Reply by RI-ss** [reply]
change the clang-cl to clang ,clangd could find iostream and sdl2 correctly
![image](https://user-images.githubusercontent.com/71321125/219857533-94cf8d42-0d7d-44d3-b941-85192fa89d87.png)
![image](https://user-images.githubusercontent.com/71321125/219857542-b3545a95-4b39-4f97-85f8-506369e27b63.png)
![image](https://user-images.githubusercontent.com/71321125/219857553-9d3204d3-26df-4b8d-ad8b-ca139e8f5af8.png)
but the build tool cannt build successfully through clang,it would be ok in clang-cl
![image](https://user-images.githubusercontent.com/71321125/219857811-d9948e7b-17db-4331-b110-04a84dea0312.png)

**Reply by HighCommander4** [reply]
Yeah, altering the `compile_commands.json` file like this feels more like a workaround than a proper fix.

In clangd 16 (scheduled to be released in a few weeks), you will be able to leave the `compile_commands.json` file unmodified (i.e. using `clang-cl.exe`) and override the compiler just for clangd using clangd's [config file](https://clangd.llvm.org/config.html), with something like:

```yaml
CompileFlags:
  Compiler: C:\\msys64\\mingw64\\bin\\clang.exe
```

However, I think even better would be to make clangd's system include extractor just work out of the box with `clang-cl.exe`.
